README.rst
Wallpaper |build-status|
========================

|logo|

This project generates (various) wallpapers using `Python <https://www.python.org/>`_
and `Pillow <https://python-pillow.org/>`_. The  default colorscheme is the `Colorbrewer <http://colorbrewer2.org/>`_
11-class diverging BrBG palette (supplied by `Palettable <https://jiffyclub.github.io/palettable/>`_).

Run ``wallpaper`` in your console to get a PNG image with the name ``wallpaper-<current-time>.png``.

Python 2 and 3 are supported.

Quick Start
===========

Install and run::

    $ pip install wallpaper
    $ wallpaper

Usage
=====

* Creating a wallpaper

  .. code:: python

      from wallpaper import Cubic

      image = Cubic()
      image.paint()

* Implementing your own

  Inherit from ``wallpaper.Wallpaper`` and implement the ``paint_pattern`` function.

TODO
====

* Better color selection

.. |logo| image:: https://raw.githubusercontent.com/mitakas/wallpaper/master/docs/wallpaper.png
    :alt: Logo

.. |build-status| image:: https://travis-ci.org/mitakas/wallpaper.svg?branch=master
    :target: https://travis-ci.org/mitakas/wallpaper
    :alt: Build status


HISTORY.rst
History
=======

0.2.3.dev0 (unreleased)
-----------------------

* fix the image in the documentation
* include the Wallpaper-class for easier use
* handle relative paths better, e.g. ~/Desktop/wallpaper.png

0.2.2 (2016-03-01)
------------------

* add more arguments to the console script

0.2.1 (2016-03-01)
------------------

* fix console script

0.2 (2016-03-01)
----------------

* create the wallpapers in a neater way, use inheritance

0.1 (2016-02-29)
----------------

* initial release


tests/test_cubic.py
from wallpaper import Cubic

class TestCubic:
    def test_parameters(self):
        x = Cubic()
        assert x.width == 1600
        assert x.height == 900
        assert x.cube_size == 50
        assert x.filename == 'wallpaper.png'


tests/test_wallpaper.py
from PIL import Image
from wallpaper import wallpaper

class TestWallpaper:
    def test_parameters(self):
        x = wallpaper.Wallpaper()
        assert x.width == 1600
        assert x.height == 900
        assert x.filename == 'wallpaper.png'

    # def test_paint(self, tmpdir):
        # tempfile = tmpdir.mkdir("sub").join('wallpaper.png')
        # x = wallpaper.Wallpaper()
        # x.paint()

        # im = Image.open(fp=tempfile)
        # color = im.getpixel(xy=(0, 0))
        # assert color == (47, 98, 135)


wallpaper/cubic.py
import random

from .wallpaper import Wallpaper

from PIL import ImageColor
from PIL import ImageDraw

from palettable.colorbrewer.diverging import BrBG_11


class Cubic(Wallpaper):
    """
    Create a wallpaper consisting of cubes.

    Parameters
    ----------
    width: int
    height: int
    cube_size: int
        Split the image into cubes with some color.
    filename: str

    Attributes
    ----------
    width: int
    height: int
    cube_size: int
        Split the image into cubes with some color.
    filename: str
    image: PIL Image
        The image which holds the wallpaper.

    """
    def __init__(self, width=1600, height=900, filename='wallpaper.png',
                 cube_size=50):
        self.cube_size = cube_size
        super(Cubic, self).__init__(width=width, height=height,
                                    filename=filename)

    def next_color(self):
        """
        Returns the next color. Currently returns a random
        color from the Colorbrewer 11-class diverging BrBG palette.

        Returns
        -------
        next_rgb_color: tuple of ImageColor

        """
        next_rgb_color = ImageColor.getrgb(random.choice(BrBG_11.hex_colors))
        return next_rgb_color

    def paint_cube(self, x, y):
        """
        Paints a cube at a certain position a color.

        Parameters
        ----------
        x: int
            Horizontal position of the upper left corner of the cube.
        y: int
            Vertical position of the upper left corner of the cube.

        """
        # get the color
        color = self.next_color()
        # calculate the position
        cube_pos = [x, y, x + self.cube_size, y + self.cube_size]
        # draw the cube
        draw = ImageDraw.Draw(im=self.image)
        draw.rectangle(xy=cube_pos, fill=color)

    def paint_pattern(self):
        """
        Paints all the cubes.

        """
        x = 0
        while x < self.width:
            y = 0
            while y < self.height:
                self.paint_cube(x, y)
                y += self.cube_size
            x += self.cube_size


wallpaper/console_script.py
import argparse
import os
import time

from . import Cubic
from . import __version__


def create_wallpaper():
    current_time = time.strftime('%Y-%m-%d-%H-%M-%S')

    parser = argparse.ArgumentParser(prog='wallpaper')
    parser.add_argument('--version', action='version', version='%(prog)s {}'.format(__version__))

    parser.add_argument('--width',
                        default=1600, type=int,
                        help='Set the width of the wallpaper.')
    parser.add_argument('--height',
                        default=900, type=int,
                        help='Set the height of the wallpaper.')
    parser.add_argument('--cube-size',
                        default=50, type=int,
                        help='Choose a size of the cube. Use a common denominator for width and height.')
    parser.add_argument('--filename', type=str,
                        default='wallpaper-{}.png'.format(current_time),
                        help='Specify the filename.')

    args = parser.parse_args()

    # workaround for ~ in the path
    args.filename = os.path.expanduser(args.filename)

    w = Cubic(width=args.width, height=args.height, filename=args.filename,
              cube_size=args.cube_size)
    w.paint()


create_wallpaper()


wallpaper/wallpaper.py
from PIL import Image


class Wallpaper(object):
    """
    Base class for all wallpapers.

    Creates a (very) simple single-color wallpaper.

    Parameters
    ----------
    width: int
    height: int
    filename: str

    Attributes
    ----------
    width: int
    height: int
    filename: str
    image: PIL Image
        The image which holds the wallpaper.

    """
    def __init__(self, width=1600, height=900, filename='wallpaper.png'):
        self.width = width
        self.height = height
        self.filename = filename

    def paint_pattern(self):
        """
        Paints the pattern. The default is just a random color.

        """
        pass

    def paint(self):
        """
        Saves the wallpaper as the specified filename.

        """
        # nice blue color
        self.image = Image.new(mode='RGB', size=(self.width, self.height),
                               color=(47, 98, 135))
        self.paint_pattern()
        self.image.save(fp=self.filename)
